core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -733,8 +733,8 @@ private void copyLogSegment(UnifiedLog log, LogSegment segment, long nextSegment
             remoteLogMetadataManager.addRemoteLogSegmentMetadata(copySegmentStartedRlsm).get();
 
             ByteBuffer leaderEpochsIndex = getLeaderEpochCheckpoint(log, -1, nextSegmentBaseOffset).readAsByteBuffer();
-            LogSegmentData segmentData = new LogSegmentData(logFile.toPath(), toPathIfExists(segment.lazyOffsetIndex().get().file()),
-                    toPathIfExists(segment.lazyTimeIndex().get().file()), Optional.ofNullable(toPathIfExists(segment.txnIndex().file())),
+            LogSegmentData segmentData = new LogSegmentData(logFile.toPath(), toPathIfExists(segment.offsetIndex().file()),
+                    toPathIfExists(segment.timeIndex().file()), Optional.ofNullable(toPathIfExists(segment.txnIndex().file())),
```

Comment:
@satishd Is it intentional that we force the indexes to be materialized here? We could pass the file without materializing if that's better.

Reply:
This metadata is passed to RSM plugin which is external to Kafka. I would like to hide the details (that we have a lazy materialized index) from the external RSM plugin and instead have a clean contract which states, "Kafka guarantees that these files will be present, RSM can pick up and upload these". This provides a clean decoupling where Kafka <-> RSM plugin state sharing is only via files.
That is why we need to materialize the indexes before giving them to RSM plugin.

Reply:
There is cost in materializing and hence why we should be sure it's needed. If that's the case here, then we're all good.

Reply:
I have been thinking about this. I take back my original statement.

We want to ensure that the file we are passing to RSM plugin contains all the data which is present in MemoryByteBuffer i.e. we should have flushed the MemoryByteBuffer to the file using force(). In Kafka, when we close a segment, indexes are flushed asynchronously. Hence, it might be possible that when we are passing the file to RSM, the file doesn't contain flushed data. This is a bug but it is not related to the change we are trying to make here. I will create a separate JIRA for this.

RSM doesn't need to read the index file during archive, hence it's ok to pass just the file name "without" materializing the index in the in-memory mmapped buffers.

Hence, we should flush() the content of indexes into the file before this operation but it is not necessary to materialize (i.e. read the content of file into memory).

@satishd we require your opinion here.

Reply:
Perhaps we can file the JIRA and discuss it there.